scipy/spatial/transform/rotation.py

```diff
@@ -8,6 +8,94 @@
 AXIS_TO_IND = {'x': 0, 'y': 1, 'z': 2}
 
 
+def _elementary_basis_vector(char):
+    b = np.zeros(3)
+    b[AXIS_TO_IND[char]] = 1
+    return b
+
+
+def _make_euler_from_dcm(dcm, seq, extrinsic=False):
+    # The algorithm assumes intrinsic frame transformations. Their
+    # ELEMENTARY dcms are transpose of what we return.
+    # Adapt the algorithm for our case by
+    # 1. Transposing our dcm representation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+
+    seq = seq[::-1] if extrinsic else seq
+
+    if dcm.ndim == 2:
+        dcm = dcm[None, :, :]
+    num_rotations = dcm.shape[0]
+
+    # Step 0
+    # Algorithms dcms are transpose of ours
+    dcm = np.transpose(dcm, (0, 2, 1))
+    # Algorithm assumes axes as column vectors, here we use 1D vectors
+    n1 = _elementary_basis_vector(seq[0])
+    n2 = _elementary_basis_vector(seq[1])
+    n3 = _elementary_basis_vector(seq[2])
+
+    # Step 2
+    lamb = np.arctan2(np.dot(np.cross(n1, n2), n3), np.dot(n1, n3))
+    c = np.empty((3, 3))
+    c[0] = n2
+    c[1] = np.cross(n1, n2)
+    c[2] = n1
+
+    # Step 3
+    cl = np.cos(lamb)
+    sl = np.sin(lamb)
+    rt = np.array([
+        [1, 0, 0],
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    ct = c.T
+    o = np.empty_like(dcm)
+    for ind in range(num_rotations):
+        o[ind] = rtc.dot(dcm[ind]).dot(ct)
+
+    # Step 4
+    angle2 = lamb + np.arccos(o[:, 2, 2])
+
+    # Steps 5, 6
+    eps = np.finfo(float).resolution  # ~1e-15
+    safe1 = (np.abs(angle2 - lamb) >= eps)
+    safe2 = (np.abs(angle2 - lamb - np.pi) >= eps)
+
+    angle1 = np.empty(num_rotations)
+    angle3 = np.empty(num_rotations)
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angle1[safe_mask] = np.arctan2(o[safe_mask, 2, 0], -o[safe_mask, 2, 1])
+    angle3[safe_mask] = np.arctan2(o[safe_mask, 0, 2], o[safe_mask, 1, 2])
+
+    # 6a
+    angle3[~safe_mask] = 0
+    # 6b
+    angle1[~safe1] = np.arctan2(o[~safe1, 0, 1] - o[~safe1, 1, 0],
+                                o[~safe1, 0, 0] + o[~safe1, 1, 1])
+    # 6c
+    angle1[~safe2] = np.arctan2(o[~safe2, 0, 1] + o[~safe2, 1, 0],
+                                o[~safe2, 0, 0] - o[~safe2, 1, 1])
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    adjust_mask = np.logical_or(angle2 < 0, angle2 > np.pi)
+    angle1[adjust_mask] = (angle1[adjust_mask] + np.pi) % (2 * np.pi)
+    angle2[adjust_mask] = (2 * lamb - angle2[adjust_mask]) % (2 * np.pi)
+    angle3[adjust_mask] = (angle3[adjust_mask] - np.pi) % (2 * np.pi)
+
+    # Step 8
+    # TODO: if any observability flags are poor, possibly raise a UserWarning?
```

Comment:
I believe it is a good idea. Here we have this notorious "gimbal lock", although I don't like this wording. We should say that we can't determine the angles uniquely and the third angle was set to 0 for some rotations. 

But we need to be careful, it might be not just 0 because of the adjustments coded. Needs some thinking.